Toolchain version is not supported.

### Describe the bug.

receive this error ether on macos and on windows 

after this 

git clone --recursive https://github.com/flipperdevices/flipperzero-firmware.git

when I try to compile with ./fbt

get this all the time

scons: *** Toolchain version is not supported. Allowed: (' 10.3.',), toolchain:    Stop.

### Reproduction

git clone --recursive https://github.com/flipperdevices/flipperzero-firmware.git

./fbt


### Target

_No response_

### Logs

_No response_

### Anything else?

_No response_

@unixispower This fixed the issue for me. huge thanks. But now I get the error:         


LINK    build\f7-firmware-C\firmware.elf
g:/flipper_zero/unleashed-firmware/toolchain/x86_64-windows/bin/../lib/gcc/arm-n
one-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: build/f7-firmware-C/lib\li
bflipper7.a(furi_hal_bt.o): in function `furi_hal_bt_get_conn_rssi':
G:\flipper_zero\unleashed-firmware/firmware\targets\f7\furi_hal/furi_hal_bt.c:47
7: undefined reference to `gap_get_remote_conn_rssi'
collect2.exe: error: ld returned 1 exit status
scons: *** [build\f7-firmware-C\firmware.elf] Error 1

********** FBT ERRORS **********
build\f7-firmware-C\firmware.elf: Error 1

--- a/scripts/fbt_tools/compilation_db.py
+++ b/scripts/fbt_tools/compilation_db.py
@@ -32,7 +32,7 @@ which is the name that most clang tools search for by default.
 import fnmatch
 import itertools
 import json
-from functools import lru_cache
+from shlex import quote
 
 import SCons
 from SCons.Tool.asm import ASPPSuffixes, ASSuffixes
@@ -45,6 +45,8 @@ from SCons.Tool.cxx import CXXSuffixes
 # access to write all of the files. But it seems clunky. How can the emitter and the scanner
 # communicate more gracefully?
 __COMPILATION_DB_ENTRIES = []
+
+# We cache the tool path lookups to avoid doing them over and over again.
 _TOOL_PATH_CACHE = {}
 
 
@@ -123,6 +125,14 @@ def compilation_db_entry_action(target, source, env, **kw):
         source=env["__COMPILATIONDB_USOURCE"],
         env=env["__COMPILATIONDB_ENV"],
     )
+
+    # We assume first non-space character is the executable
+    executable = command.split(" ", 1)[0]
+    if not (tool_path := _TOOL_PATH_CACHE.get(executable, None)):
+        tool_path = env.WhereIs(executable) or executable
+        _TOOL_PATH_CACHE[executable] = tool_path
+    # Replacing the executable with the full path
+    command = quote(tool_path) + command[len(executable) :]
 
     entry = {
         "directory": env.Dir("#").abspath,
@@ -244,19 +254,13 @@ def generate(env, **kwargs):
     for entry in components_by_suffix:
         suffix = entry[0]
         builder, base_emitter, command = entry[1]
-        if not (tool_path := _TOOL_PATH_CACHE.get(command, None)):
-            tool_path = env.WhereIs(command) or command
-            _TOOL_PATH_CACHE[command] = tool_path
-
-        # Assumes a dictionary emitter
-        emitter = builder.emitter.get(suffix, False)
-        if emitter:
+        if emitter := builder.emitter.get(suffix, False):
             # We may not have tools installed which initialize all or any of
             # cxx, cc, or assembly. If not skip resetting the respective emitter.
             builder.emitter[suffix] = SCons.Builder.ListEmitter(
                 [
                     emitter,
-                    make_emit_compilation_DB_entry(tool_path),
+                    make_emit_compilation_DB_entry(command),
                 ]
             )
 

--- a/scripts/fbt_tools/compilation_db.py
+++ b/scripts/fbt_tools/compilation_db.py
@@ -29,21 +29,23 @@ which is the name that most clang tools search for by default.
 # WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 #
 
-import json
-import itertools
 import fnmatch
+import itertools
+import json
+from functools import lru_cache
+
 import SCons
-
-from SCons.Tool.cxx import CXXSuffixes
+from SCons.Tool.asm import ASPPSuffixes, ASSuffixes
 from SCons.Tool.cc import CSuffixes
-from SCons.Tool.asm import ASSuffixes, ASPPSuffixes
+from SCons.Tool.cxx import CXXSuffixes
 
-# TODO FL-3542: Is there a better way to do this than this global? Right now this exists so that the
+# TODO: (-nofl) Is there a better way to do this than this global? Right now this exists so that the
 # emitter we add can record all of the things it emits, so that the scanner for the top level
 # compilation database can access the complete list, and also so that the writer has easy
 # access to write all of the files. But it seems clunky. How can the emitter and the scanner
 # communicate more gracefully?
 __COMPILATION_DB_ENTRIES = []
+_TOOL_PATH_CACHE = {}
 
 
 # We make no effort to avoid rebuilding the entries. Someday, perhaps we could and even
@@ -91,7 +93,7 @@ def make_emit_compilation_DB_entry(comstr):
             __COMPILATIONDB_ENV=env,
         )
 
-        # TODO FL-3541: Technically, these next two lines should not be required: it should be fine to
+        # TODO: (-nofl) Technically, these next two lines should not be required: it should be fine to
         # cache the entries. However, they don't seem to update properly. Since they are quick
         # to re-generate disable caching and sidestep this problem.
         env.AlwaysBuild(entry)
@@ -242,6 +244,9 @@ def generate(env, **kwargs):
     for entry in components_by_suffix:
         suffix = entry[0]
         builder, base_emitter, command = entry[1]
+        if not (tool_path := _TOOL_PATH_CACHE.get(command, None)):
+            tool_path = env.WhereIs(command) or command
+            _TOOL_PATH_CACHE[command] = tool_path
 
         # Assumes a dictionary emitter
         emitter = builder.emitter.get(suffix, False)
@@ -251,7 +256,7 @@ def generate(env, **kwargs):
             builder.emitter[suffix] = SCons.Builder.ListEmitter(
                 [
                     emitter,
-                    make_emit_compilation_DB_entry(command),
+                    make_emit_compilation_DB_entry(tool_path),
                 ]
             )
 

--- a/scripts/fbt_tools/compilation_db.py
+++ b/scripts/fbt_tools/compilation_db.py
@@ -135,7 +135,7 @@ def compilation_db_entry_action(target, source, env, **kw):
     if " " in tool_path:
         tool_path = quote(tool_path)
     # Replacing the executable with the full path
-    command = quote(tool_path) + command[len(executable) :]
+    command = tool_path + command[len(executable) :]
 
     entry = {
         "directory": env.Dir("#").abspath,

--- a/scripts/fbt_tools/compilation_db.py
+++ b/scripts/fbt_tools/compilation_db.py
@@ -131,6 +131,9 @@ def compilation_db_entry_action(target, source, env, **kw):
     if not (tool_path := _TOOL_PATH_CACHE.get(executable, None)):
         tool_path = env.WhereIs(executable) or executable
         _TOOL_PATH_CACHE[executable] = tool_path
+    # If there are spaces in the executable path, we need to quote it
+    if " " in tool_path:
+        tool_path = quote(tool_path)
     # Replacing the executable with the full path
     command = quote(tool_path) + command[len(executable) :]
 

--- a/scripts/fbt_tools/crosscc.py
+++ b/scripts/fbt_tools/crosscc.py
@@ -2,7 +2,6 @@ import subprocess
 
 import gdb
 import objdump
-import shutil
 
 import strip
 from SCons.Action import _subproc
@@ -13,7 +12,7 @@ from SCons.Tool import ar, asm, gcc, gnulink, gxx
 def prefix_commands(env, command_prefix, cmd_list):
     for command in cmd_list:
         if command in env:
-            env[command] = shutil.which(command_prefix + env[command])
+            env.Replace(**{command: command_prefix + env[command]})
 
 
 def _get_tool_version(env, tool):

--- a/scripts/fbt_tools/crosscc.py
+++ b/scripts/fbt_tools/crosscc.py
@@ -2,7 +2,6 @@ import subprocess
 
 import gdb
 import objdump
-
 import strip
 from SCons.Action import _subproc
 from SCons.Errors import StopError
@@ -12,20 +11,25 @@ from SCons.Tool import ar, asm, gcc, gnulink, gxx
 def prefix_commands(env, command_prefix, cmd_list):
     for command in cmd_list:
         if command in env:
-            env.Replace(**{command: command_prefix + env[command]})
+            prefixed_binary = command_prefix + env[command]
+            if not env.WhereIs(prefixed_binary):
+                raise StopError(
+                    f"Toolchain binary {prefixed_binary} not found in PATH."
+                )
+            env.Replace(**{command: prefixed_binary})
 
 
 def _get_tool_version(env, tool):
     verstr = "version unknown"
     proc = _subproc(
         env,
-        env.subst("${%s} --version" % tool),
+        [env.subst("${%s}" % tool), "--version"],
         stdout=subprocess.PIPE,
         stderr="devnull",
         stdin="devnull",
         universal_newlines=True,
         error="raise",
-        shell=True,
+        shell=False,
     )
     if proc:
         verstr = proc.stdout.readline()

--- a/scripts/fbt_tools/pvsstudio.py
+++ b/scripts/fbt_tools/pvsstudio.py
@@ -48,6 +48,7 @@ def generate(env):
             "@.pvsoptions",
             "-j${PVSNCORES}",
             # "--incremental", # kinda broken on PVS side
+            "--disableLicenseExpirationCheck",
         ],
         PVSCONVOPTIONS=[
             "-a",

